### PR TITLE
replace `DefaultExecutionConfig` with `ExecutionConfig()`

### PR DESCRIPTION
**Pull request by andrijapau**
**Context:**

https://github.com/PennyLaneAI/pennylane/pull/7697 changed the default usage of `ExecutionConfig` to not be `DefaultExecutionConfig` but instead just creating a new instance of `ExecutionConfig`.

**Description of the Change:**

As outlined above.

**Benefits:** Better code.

**Possible Drawbacks:** None identified.

[sc-93591]

### CODE REVIEW

File: pennylane_lightning/lightning_base/lightning_base.py [hunk 1]
```diff
@@ -159,7 +159,7 @@ def dynamic_wires_from_circuit(self, circuit):
         return circuit
 
     @abstractmethod
-    def preprocess(self, execution_config: ExecutionConfig = DefaultExecutionConfig):
+    def preprocess(self, execution_config: ExecutionConfig = None):
```
Comment (albi3ro) [inline pennylane_lightning/lightning_base/lightning_base.py]:
```suggestion
    def preprocess(self, execution_config: Optional[ExecutionConfig] = None):
```

File: pennylane_lightning/lightning_base/lightning_base.py [hunk 2]
```diff
@@ -183,7 +183,7 @@ def preprocess(self, execution_config: ExecutionConfig = DefaultExecutionConfig)
     def execute(
         self,
         circuits: QuantumTape_or_Batch,
-        execution_config: ExecutionConfig = DefaultExecutionConfig,
+        execution_config: ExecutionConfig = None,
```
Comment (albi3ro) [inline pennylane_lightning/lightning_base/lightning_base.py]:
```suggestion
        execution_config: Optional[ExecutionConfig] = None,
```

**Comment by andrijapau** [comment]
@maliasadi , do you have capacity to review this PR? Also look likes code cov is complaining in correctly again.

File: pennylane_lightning/lightning_base/lightning_base.py [hunk 3]
```diff
@@ -403,17 +403,20 @@ def simulate_and_vjp(  # pylint: disable=too-many-arguments, too-many-positional
     def compute_derivatives(
         self,
         circuits: QuantumTape_or_Batch,
-        execution_config: ExecutionConfig = DefaultExecutionConfig,
+        execution_config: ExecutionConfig = None,
```
Comment (albi3ro) [inline pennylane_lightning/lightning_base/lightning_base.py]:
```suggestion
        execution_config: Optional[ExecutionConfig] = None,
```
May need to do a pass to all the other methods for the same change.

Comment (andrijapau) [reply pennylane_lightning/lightning_base/lightning_base.py]: 👍🏼

Comment (andrijapau) [reply pennylane_lightning/lightning_base/lightning_base.py]: https://github.com/PennyLaneAI/pennylane-lightning/pull/1185/commits/ae1a4837bf8d324ae0799ab58de873c5c1e39417

File: tests/lightning_base/test_device.py [hunk 4]
```diff
@@ -543,7 +543,7 @@ def process_and_execute(device, tape):
         "config, expected_config",
         [
             (
-                DefaultExecutionConfig,
+                ExecutionConfig(),
```
Comment (albi3ro) [inline tests/lightning_base/test_device.py:546]: Might actually be worth passing in `None` here. Might help on code coverage.

Comment (albi3ro) [reply tests/lightning_base/test_device.py:546]: Same with the similar tests.

Comment (andrijapau) [inline tests/lightning_base/test_device.py:546]:
```suggestion
                None,
```

File: tests/lightning_base/test_device.py [hunk 5]
```diff
@@ -1016,7 +1016,7 @@ def process_and_execute(device, tape, execute_and_derivatives=False, obs_batch=F
         "config, tape, expected",
         [
             (None, None, True),
-            (DefaultExecutionConfig, None, False),
+            (ExecutionConfig(), None, False),
```
Comment (andrijapau) [inline tests/lightning_base/test_device.py:1033]:
```suggestion
            (None, None, False),
```

File: tests/lightning_base/test_device.py [hunk 6]
```diff
@@ -1395,7 +1395,7 @@ def process_and_execute(device, tape, dy, execute_and_derivatives=False, obs_bat
...
```
Comment (andrijapau) [inline tests/lightning_base/test_device.py:1422]:
```suggestion
            (None, None, False),
```

File: tests/lightning_base/test_simulate_method.py [hunk 7]
```diff
@@ -105,7 +105,7 @@ def test_sample_values_with_mcmc(self, lightning_sv, tol, mcmc, kernel):
             "num_burnin": 100,
         }
 
-        execution_config = DefaultExecutionConfig
+        execution_config = ExecutionConfig()
```
Comment (andrijapau) [inline tests/lightning_base/test_simulate_method.py:108]:
```suggestion
        execution_config = None
```

File: tests/test_jaxpr_jvp.py [hunk 8]
```diff
@@ -52,7 +52,7 @@ def circuit(x):
         args = (0.5,)
         jaxpr = jax.make_jaxpr(circuit)(*args)
 
-        execution_config = DefaultExecutionConfig
+        execution_config = ExecutionConfig()
```
Comment (andrijapau) [inline tests/test_jaxpr_jvp.py]:
```suggestion
        execution_config = None
```

File: tests/lightning_base/test_device.py (hunk 4 above)
Comment (andrijapau) [reply tests/lightning_base/test_device.py:546]: Done https://github.com/PennyLaneAI/pennylane-lightning/pull/1185/commits/4626c4c7a29d1a9168b726517abf371693288ffa

File: tests/test_jaxpr_jvp.py [hunk 9]
```diff
@@ -52,7 +52,7 @@ def circuit(x):
         args = (0.5,)
         jaxpr = jax.make_jaxpr(circuit)(*args)
 
-        execution_config = DefaultExecutionConfig
+        execution_config = None
         execution_config.gradient_method = "backprop"
```
Comment (albi3ro) [inline tests/test_jaxpr_jvp.py]: 🤦‍♀️ 

```suggestion
        execution_config = qml.devices.ExecutionConfig(gradient_method="backprop")
```

File: tests/lightning_base/test_device.py [hunk 10]
```diff
@@ -1016,7 +1016,7 @@ def process_and_execute(device, tape, execute_and_derivatives=False, obs_batch=F
         "config, tape, expected",
         [
             (None, None, True),
-            (DefaultExecutionConfig, None, False),
+            (None, None, False),
```
Comment (andrijapau) [inline tests/lightning_base/test_device.py]:
```suggestion
            (ExecutionConfig(), None, False),
```

File: tests/lightning_base/test_device.py [hunk 11]
```diff
@@ -1395,7 +1395,7 @@ def process_and_execute(device, tape, dy, execute_and_derivatives=False, obs_bat
...
```
Comment (andrijapau) [inline tests/lightning_base/test_device.py]:
```suggestion
            (ExecutionConfig(), None, False),
```

File: pennylane_lightning/lightning_base/lightning_base.py [hunk 12]
```diff
@@ -403,17 +403,20 @@ def simulate_and_vjp(  # pylint: disable=too-many-arguments, too-many-positional
...
         circuits: QuantumTape_or_Batch,
-        execution_config: ExecutionConfig = DefaultExecutionConfig,
+        execution_config: Optional[ExecutionConfig] = None,
     ) -> Tuple:
         """Calculate the jacobian of either a single or a batch of circuits on the device.
 
         Args:
             circuits (Union[QuantumTape, Sequence[QuantumTape]]): the circuits to calculate derivatives for
-            execution_config (ExecutionConfig): a data structure with all additional information required for execution. Default is ``DefaultExecutionConfig``.
+            execution_config (ExecutionConfig): a data structure with all additional information required for execution. Default is ``None``.
```
Comment (maliasadi) [inline pennylane_lightning/lightning_base/lightning_base.py]:
```suggestion
            execution_config (ExecutionConfig): a data structure with all additional information required for execution. Default is ``None``, which sets the execution config to the default setup.
```

File: pennylane_lightning/lightning_base/lightning_base.py [hunk 13]
```diff
@@ -429,17 +432,20 @@ def compute_derivatives(
...
         circuits: QuantumTape_or_Batch,
-        execution_config: ExecutionConfig = DefaultExecutionConfig,
+        execution_config: Optional[ExecutionConfig] = None,
     ) -> Tuple:
         """Compute the results and jacobians of circuits at the same time.
 
         Args:
             circuits (Union[QuantumTape, Sequence[QuantumTape]]): the circuits or batch of circuits
-            execution_config (ExecutionConfig): a data structure with all additional information required for execution. Default is ``DefaultExecutionConfig``.
+            execution_config (ExecutionConfig): a data structure with all additional information required for execution. Default is ``None``.
```
Comment (maliasadi) [inline pennylane_lightning/lightning_base/lightning_base.py]:
```suggestion
            execution_config (ExecutionConfig): a data structure with all additional information required for execution. Default is ``None``, which sets the execution config to the default setup.
```

File: pennylane_lightning/lightning_base/lightning_base.py [hunk 14]
```diff
@@ -471,7 +477,7 @@ def compute_vjp(
         self,
         circuits: QuantumTape_or_Batch,
         cotangents: Tuple[Number],
-        execution_config: ExecutionConfig = DefaultExecutionConfig,
+        execution_config: Optional[ExecutionConfig] = None,
```
Comment (maliasadi) [inline pennylane_lightning/lightning_base/lightning_base.py]: Would you update the docstring for the default `execution_config`?

File: .github/CHANGELOG.md [hunk 15]
```diff
@@ -52,6 +52,9 @@
   [(#1175)](https://github.com/PennyLaneAI/pennylane-lightning/pull/1175)
 
 <h3>Internal changes ⚙️</h3>
+
+- Replace `DefaultExecutionConfig` with `ExecutionConfig()`.
```
Comment (maliasadi) [inline .github/CHANGELOG.md]:
```suggestion
- Replace `DefaultExecutionConfig` with `ExecutionConfig()` across the Lightning Python device APIs.
```

**Review by maliasadi — COMMENTED** [review]
Thanks @andrijapau! Happy to approve after checking the wheels (`ci:build-wheels`) and GPU devices (`ci:use-gpu-runner`)

File: pennylane_lightning/lightning_base/lightning_base.py (hunk 14 above)
Comment (andrijapau) [reply pennylane_lightning/lightning_base/lightning_base.py]: https://github.com/PennyLaneAI/pennylane-lightning/pull/1185/commits/652b420ff30a223a7a8b776e20ff2a570cb4dbfc Done!

**Comment by andrijapau** [comment]
@maliasadi , where should I add tests for the remaining coverage? 🫠

**Comment by LuisAlfredoNu** [comment]
Hi @andrijapau  
I will add the label `do not merge` for Release purposes. Feel free to remove it after the `v0.42.0` version has been released  
:grin:

File: pennylane_lightning/lightning_base/lightning_base.py [hunk 16]
```diff
@@ -632,7 +645,7 @@ def jaxpr_jvp(
         jaxpr: "jax.extend.core.Jaxpr",
         args: Sequence[TensorLike],
         tangents: Sequence[TensorLike],
-        execution_config: Optional[ExecutionConfig] = None,
+        execution_config: ExecutionConfig | None = None,
```
Comment (LuisAlfredoNu) [inline pennylane_lightning/lightning_base/lightning_base.py:648]: This section doesn't need the code 
```
        if execution_config is None:
            execution_config = ExecutionConfig()

```
?

Comment (andrijapau) [reply pennylane_lightning/lightning_base/lightning_base.py:648]: That kwarg isn't being used in the function.

File: tests/lightning_base/test_device.py [hunk 17]
```diff
@@ -543,7 +543,7 @@ def process_and_execute(device, tape):
         "config, expected_config",
         [
             (
-                DefaultExecutionConfig,
+                None,
```
Comment (LuisAlfredoNu) [inline tests/lightning_base/test_device.py:555]: Could you add an additional test with `ExecutionConfig()`instead of `None`, please?

File: tests/lightning_base/test_device.py (hunk 6 above)
Comment (LuisAlfredoNu) [reply tests/lightning_base/test_device.py:1422]: That test will be useful

File: tests/lightning_base/test_device.py (hunk 17 above)
Comment (andrijapau) [reply tests/lightning_base/test_device.py:555]: https://github.com/PennyLaneAI/pennylane-lightning/pull/1185/commits/58d80dff1862502d7e245cfa5970ef5952820bd9

File: tests/lightning_base/test_device.py (hunk 6 above)
Comment (andrijapau) [reply tests/lightning_base/test_device.py:1422]: Which?

File: tests/test_jaxpr_jvp.py [hunk 18]
```diff
@@ -52,8 +52,7 @@ def circuit(x):
         args = (0.5,)
         jaxpr = jax.make_jaxpr(circuit)(*args)
 
-        execution_config = DefaultExecutionConfig
-        execution_config.gradient_method = "backprop"
+        execution_config = qml.devices.ExecutionConfig(gradient_method="backprop")
```
Comment (LuisAlfredoNu) [inline tests/test_jaxpr_jvp.py]: If you will use the full call stating with `qml....` then you need to remove the import in top of the file

File: tests/lightning_base/test_device.py (hunk 6 above)
Comment (LuisAlfredoNu) [reply tests/lightning_base/test_device.py:1422]: yep, both test :smile:

File: tests/test_jaxpr_jvp.py (hunk 18 above)
Comment (andrijapau) [reply tests/test_jaxpr_jvp.py]: https://github.com/PennyLaneAI/pennylane-lightning/pull/1185/commits/f03eb117edec690f874d302580da457066966cd7

Comment (andrijapau) [reply tests/test_jaxpr_jvp.py]: Good catch :)

File: tests/lightning_base/test_device.py (hunk 6 above)
Comment (andrijapau) [reply tests/lightning_base/test_device.py:1422]: Hmm. How do we want the logic for supports_derivatives? Right now it's this,
```python
        if execution_config is None:
            return True

        if execution_config.gradient_method in {"adjoint", "best"}:
            if circuit is None:
                return True
            return _supports_adjoint(circuit=circuit)

        return False
```
but it was,
```python
        if execution_config is None and circuit is None:
            return True
        if execution_config.gradient_method not in {"adjoint", "best"}:
            return False
        if circuit is None:
            return True
        return _supports_adjoint(circuit=circuit)
```

File: pennylane_lightning/lightning_base/lightning_base.py (hunk 16 above)
Comment (LuisAlfredoNu) [reply pennylane_lightning/lightning_base/lightning_base.py:648]: L690

Comment (andrijapau) [reply pennylane_lightning/lightning_base/lightning_base.py:648]: Ah, I was looking at eval_jaxpr

Comment (andrijapau) [reply pennylane_lightning/lightning_base/lightning_base.py:648]: Done https://github.com/PennyLaneAI/pennylane-lightning/pull/1185/commits/adfb1cf954b02b93c5d39977c01e53785bb26ba6

File: pennylane_lightning/lightning_gpu/lightning_gpu.py [hunk 19]
```diff
@@ -449,13 +455,15 @@ def supports_derivatives(
             Bool: Whether or not a derivative can be calculated provided the given information
 
         """
-        if execution_config is None and circuit is None:
-            return True
-        if execution_config.gradient_method not in {"adjoint", "best"}:
-            return False
-        if circuit is None:
+        if execution_config is None:
```
Comment (LuisAlfredoNu) [inline pennylane_lightning/lightning_gpu/lightning_gpu.py]: I think the logic is a bit different here than the original. What if `circuit != None and execution_config is None`. This statement should call `_supports_adjoint(circuit)` not just pass `True` :thinking:

File: tests/lightning_base/test_device.py (hunk 6 above)
Comment (LuisAlfredoNu) [reply tests/lightning_base/test_device.py:1422]: You are right! I think we need to change a bit the supported_derivatives

Comment (LuisAlfredoNu) [reply tests/lightning_base/test_device.py:1422]: Could you add the following tests:
```
(
    None, 
    QuantumScript([qml.RX(0.123, 0)], [qml.expval(qml.Z(0))]),
    False,
),
(
    None,
    QuantumScript([qml.RX(0.123, 0)], [qml.var(qml.Z(0))]),
    False,
)
```

Comment (andrijapau) [reply tests/lightning_base/test_device.py:1422]: Done https://github.com/PennyLaneAI/pennylane-lightning/pull/1185/commits/dae7558f08196a35fe81b937cd804fec1fe9921e

File: pennylane_lightning/lightning_gpu/lightning_gpu.py (hunk 19 above)
Comment (andrijapau) [reply pennylane_lightning/lightning_gpu/lightning_gpu.py]: I'll just revert it

File: tests/lightning_base/test_device.py (hunk 6 above)
Comment (LuisAlfredoNu) [reply tests/lightning_base/test_device.py:1422]: Do you agree that this should be False right?. becuase `ExecutionConfig` will not have the correct setup for `gradient_method=`

File: pennylane_lightning/lightning_gpu/lightning_gpu.py (hunk 19 above)
Comment (andrijapau) [reply pennylane_lightning/lightning_gpu/lightning_gpu.py]: Reverted all of them

File: tests/lightning_base/test_simulate_method.py [hunk 20]
```diff
@@ -105,7 +105,7 @@ def test_sample_values_with_mcmc(self, lightning_sv, tol, mcmc, kernel):
             "num_burnin": 100,
         }
 
-        execution_config = DefaultExecutionConfig
+        execution_config = None
```
Comment (andrijapau) [inline tests/lightning_base/test_simulate_method.py]:
```suggestion
        execution_config = ExecutionConfig()
```

File: pennylane_lightning/lightning_gpu/lightning_gpu.py [hunk 21]
```diff
@@ -451,11 +457,13 @@ def supports_derivatives(
...
-        if circuit is None:
-            return True
-        return _supports_adjoint(circuit=circuit)
+
+        if execution_config and execution_config.gradient_method in {"adjoint", "best"}:
+            if circuit is None:
+                return True
+            return _supports_adjoint(circuit=circuit)
+
+        return False
```
Comment (andrijapau) [inline pennylane_lightning/lightning_gpu/lightning_gpu.py:466]: @LuisAlfredoNu what do you think of the new logic? It follows the same pattern as what we have in default.qubit.

**Comment by andrijapau** [comment]
@albi3ro do you mind taking another look? A decent amount has changed since your last review 🙇🏼
